### PR TITLE
cloud: hide o11y for non-operator users

### PR DESCRIPTION
**Pull request by michaellzc**
part of 

- https://github.com/sourcegraph/customer/issues/1743
- https://github.com/sourcegraph/security-issues/issues/330

This is only the first part of the implementation, preventing people from finding links to `/-/debug` in the first place.

Will follow up with another one to block it from the frontend backend

## Test plan

1. Get the `cloud-site-config.json` (or use the `signed-clout-site-config` from [1password](https://start.1password.com/open/i?a=HEDEDSLHPBFGRBTKAKJWE23XX4&v=dnrhbauihkhjs5ag6vszsme45a&i=eyhez5jvjmdd745zu6jwferxha&h=team-sourcegraph.1password.com)

2. Get the signed config

  ```sh
  mi sign-cloud-site-config --config-file ./cloud-site-config.json
  ```

3. Add the following to `sg.config.overwrite.yaml`

  ```yaml
  commands:
    frontend:
      env:
        SRC_CLOUD_SITE_CONFIG: ""
  ```

4. `DEV_WEB_BUILDER=esbuild sg start`

### Customer admin cannot see o11y menu items

Login using the regular admin account, and confirm you can still see the o11y menu items:

229728069-8b4df445-242c-4744-8521-09db06c9e2a5

Create `sourcegraph-operator-site-admin-hide-maintenance` feature flag and set to `true`

Go back to site-admin, confirm you can no longer see the o11y menu items

### Operator can still see o11y menu items

Login from https://sourcegraph.test:3443/sign-in?sourcegraph-operator

Go to site-admin, confirm you can see the o11y menu items

Side-by-side (left is operator's view):

CleanShot 2023-04-04 at 16 53 27@2x

## App preview:

- [Web](https://sg-web-12-19-cloud-hide-o11y-for-non.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by michaellzc** [comment]
Current dependencies on/for this PR:
* main
  * **PR #45827** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/45827" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/45827?utm_source=stack-comment).

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (+0.11 kb) | 0.00% (+0.67 kb) | 0.00% (+0.56 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/12-19-cloud_hide_o11y_for_non-operator_users/compare-report.html;tab=live_object) for a full comparison between the commits 7c7d9ae and 40eedca or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

File: client/web/src/featureFlags/featureFlags.ts [hunk 1]
```diff
@@ -18,6 +18,7 @@ export type FeatureFlagName =
     | 'search-results-keyboard-navigation'
     | 'enable-streaming-git-blame'
     | 'plg-enable-add-codehost-widget'
+    | 'hide-observability-only-visible-to-sourcegraph-operator'
```
Comment (bobheadxi) [inline client/web/src/featureFlags/featureFlags.ts:21]:
```suggestion
    | 'sourcegraph-operator-only-observability'
```

Then all our operator-related FFs can have the same prefix

File: client/web/src/site-admin/SiteAdminArea.tsx [hunk 2]
```diff
@@ -163,12 +168,76 @@ const AuthenticatedSiteAdminArea: React.FunctionComponent<React.PropsWithChildre
 
     const [isAdminAnalyticsDisabled] = useFeatureFlag('admin-analytics-disabled', false)
 
-    const adminSideBarGroups = useMemo(() => {
-        if (isAdminAnalyticsDisabled) {
-            return props.sideBarGroups
-        }
-        return [analyticsGroup, ...props.sideBarGroups.filter(group => !isEqual(group, overviewGroup))]
```
Comment (michaellzc) [inline client/web/src/site-admin/SiteAdminArea.tsx:170]: I belive the label name of group is expected to be unique, or user will use duplicated group name on the UI which is confusing.

so we can remove the deep equal here, checking the label name should be enough.

**Review by unknwon — APPROVED** [review]
This is really stretching my TS knowledge, but LGTM 😂

File: client/web/src/featureFlags/featureFlags.ts [hunk 3]
```diff
@@ -18,6 +18,7 @@ export type FeatureFlagName =
     | 'search-results-keyboard-navigation'
     | 'enable-streaming-git-blame'
     | 'plg-enable-add-codehost-widget'
+    | 'sourcegraph-operator-only-observability'
```
Comment (bobheadxi) [inline client/web/src/featureFlags/featureFlags.ts:21]: Looking a bit closer, should we extend this flag to cover all things we want to hide under the `Maintainence` category? Other pages we can hide:

- Updates (customers have no control over this)
- Migrations (customers also have no control over this)

Then the flag might be:

```suggestion
    | 'sourcegraph-operator-site-admin-hide-maintainence'
```

**Review by bobheadxi — APPROVED** [review]

File: client/web/src/site-admin/SiteAdminArea.tsx [hunk 4]
```diff
@@ -64,9 +74,75 @@ interface SiteAdminAreaProps
...
+    const { data: externalAccounts, loading: isExternalAccountsLoading } = useUserExternalAccounts(
+        props.authenticatedUser.username
+    )
+    const [isSourcegraphOperatorSiteAdminHideMaintenance] = useFeatureFlag(
+        'sourcegraph-operator-site-admin-hide-maintenance'
+    )
+
+    const adminSideBarGroups = React.useMemo(
+        () =>
+            props.sideBarGroups.reduce((groups, group) => {
```
Comment (mrnugget) [inline client/web/src/site-admin/SiteAdminArea.tsx:97]: I'm wondering whether this code wouldn't be simpler by using `map`. It's really complex for what it's trying to achieve. You're mapping over a list of groups and if it's `(curr.header?.label === maintenanceGroupHeaderLabel)` then you modify it, depending on the 3 other conditions. I think that's a `map`.

You also don't need the `curr = group` and the scary comment here. And you can get rid of the last 5 lines in this anonymous function.

File: client/web/src/site-admin/SiteAdminArea.tsx [hunk 5]
```diff
@@ -64,9 +74,75 @@ interface SiteAdminAreaProps
...
+                        if (curr.header?.label === maintenanceGroupHeaderLabel) {
+                            curr = {
+                                ...curr,
+                                items: curr.items.filter(
+                                    item => !sourcegraphOperatorSiteAdminMaintenanceBlockItems.has(item.label)
+                                ),
+                            }
+                        }
+                    } else if (!externalAccounts.some(account => account.serviceType === 'sourcegraph-operator')) {
+                            if (curr.header?.label === maintenanceGroupHeaderLabel) {
```
Comment (mrnugget) [inline client/web/src/site-admin/SiteAdminArea.tsx:122]: You only ever do something (in both top-level branches) if this condition is true. Can't you make this the top-level condition and skip anything the rest of the code?

**Review by mrnugget — CHANGES_REQUESTED** [review]

**Comment by unknwon** [comment]
@mrnugget - thanks for the review! Updated with a much-simplified version of the code logic.

**Pull request merged**